src/prof_recent.c

```diff
@@ -517,33 +637,98 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count = prof_recent_alloc_dump_prepare_locked(tsd);
+	ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	prof_recent_t *dump_last = NULL;
 
-	emitter_json_array_kv_begin(&emitter, "recent_alloc");
-	for (prof_recent_t *n = prof_recent_alloc_begin(tsd);
-	    n != prof_recent_alloc_end(tsd);
-	    n = prof_recent_alloc_next(tsd, n)) {
-		prof_recent_alloc_dump_node(&emitter, n);
+	/*
+	 * We dump the list in segments, where each segment:
+	 * - is dumped under one critical section, and
+	 * - does not trigger buffered writer flushing,
+	 * so that flushing always takes place outside of the critical sections
+	 * and thus would not block allocations / deallocations from other
+	 * threads.
+	 */
```

Comment:
This might be a dumb question, but why do we need to hold any locks while dumping at all? The list is private at this point, since we popped it out of the global above.

Reply:
There are two locks we're holding:

`prof_recent_alloc_mtx`

The list is private only in the sense that new samples are not added to this list, but the list still holds existing samples, and if some of these samples are not freed yet, and if they're being freed during dumping, the freeing thread will need to modify the list (recording that the allocation is being freed as well as the freeing time and stack trace).

The `prof_recent_alloc_mtx` protects the list to which allocating threads add samples. When there's no dumping taking place, the freeing thread needs to acquire the `prof_recent_alloc_mtx` lock, because existing records might be being deleted, if another thread is adding new records to the list and the list has a finite limit.

In theory, while dumping, the freeing thread doesn't have to try to acquire the `prof_recent_alloc_mtx` lock, because it doesn't need to compete with allocating threads. However, there's still need for some lock that the dumping thread and the freeing thread cannot hold simultaneously; otherwise the dumping thread may dump a partial freeing stack trace. I just continue to use the `prof_recent_alloc_mtx` lock for this "some lock" for simplicity, since otherwise the freeing thread needs additional mechanism to determine whether there's another thread dumping or not.

`prof_recent_dump_mtx`

As I mentioned earlier, this lock strictly speaking is not necessary. It's only meant to permit only one thread to dump at a time. We can allow multiple threads to dump at the same time, but there can be a couple of issues:
- Even multiple threads can dump at the same time, they cannot simultaneously "check in" to the global list, because they need to compete for the `prof_recent_alloc_mtx` lock (and once they have it they'll pull the global list out). Therefore, each thread effectively only dumps the records that come in *after* the last "check in" time of another dumping thread. This is not perfectly intuitive from the application's perspective.
- If applications are not super careful, they likely would supply the same callback handle to all the dumping threads. If so, multiple concurrent dumping threads might invoke the callback handle simultaneously. Even if the logic underlying the callback handle is concurrency safe, the output would be interleaved with multiple pieces of dump content.
- (This point is implementation specific and may or may not be relevant if we decide to change the data structure.) The current implementation requires a dynamically allocated spare dummy head for each dumping thread. If we allow multiple threads to dump concurrently, then each thread needs to allocate itself, and if the allocation fails, then it's not able to dump. If we only allow one thread to dump at a time, then we can pre-allocate a spare dummy head at boot time, to make sure that all dumps will be successful.

Reply:
> However, there's still need for some lock that the dumping thread and the freeing thread cannot hold simultaneously; otherwise the dumping thread may dump a partial freeing stack trace.

Well, there can never be a partial freeing stack trace, because the stack trace is first completed and the record is then allowed to point to the stack trace. But, still, there can be possibilities of only dumping that the record is freed but not the freeing time / stack trace, or dumping the freeing time but not the stack trace, etc.

Reply:
I think probably better would be to have an `atomic_b_t has_been_freed`. When the freeing thread comes in, it can fill in the stack trace at free and atomic_store(&ptr->has_been_freed, true, ATOMIC_RELEASE);`. The dumping thread, when dumping, can write `bool freed = atomic_load(&ptr->has_been_freed, ATOMIC_ACQUIRE); if (freed) { dump_free_stack_trace(ptr); }`.

Two reasons:
- This lock is the reason for the whole backup/rewind thing in the buffered writer, right? I'd love to be able to get rid of it if we can; it vastly increases the complexity of the buffered writer for a very edge-casey situation.
- I think that the release -> dump -> reacquire strategy here still has some tail latency problems. The underlying mutex types don't guarantee fairness, and in fact on most platforms, they're biased towards the last acquirer. So in practice, a thread constantly acquiring, dropping, and reacquiring a mutex can end up looking like it was holding the mutex for the entire time anyways.

Reply:
This fencing strategy is a great one.

> I'd love to be able to get rid of it if we can; it vastly increases the complexity of the buffered writer for a very edge-casey situation.

Agree. I'd also love to keep the buffered writer simple.

Here's what I get out of your suggestion:
- Allocating thread:
  - Acquire `prof_recent_alloc_mtx`.
  - Record (with possible deletion of the earliest record).
  - Release `prof_recent_alloc_mtx`.
- Freeing thread:
  - Acquire `prof_recent_alloc_mtx`.
  - Record (with the atomic boolean the last one to write, and with an `ATOMIC_RELEASE` fence).
  - Release `prof_recent_alloc_mtx`.
- Dumping thread:
  - Acquire `prof_recent_dump_mtx`.
  - Acquire `prof_recent_alloc_mtx`.
  - Pull the entire list out.
  - Release `prof_recent_alloc_mtx`.
  - Dump (first checking the atomic boolean with an `ATOMIC_ACQUIRE` fence before dumping any deallocation information).
  - Acquire `prof_recent_alloc_mtx`.
  - Insert the list back to the start.
  - Release `prof_recent_alloc_mtx`.
  - Release `prof_recent_dump_mtx`.

Looks good?

(BTW I already have an atomic "boolean" for this purpose: the pointer on the record pointing to the allocation.)

Reply:
Yes, I think that looks great -- in particular, we moved the dumping thread from touching the hot prof_recent_alloc_mtx mutex from O(N) times to O(1).

Reply:
Well, and don't hold it during any potentially long-running operations, like I/O